src/charm.py

```diff
@@ -1,104 +1,397 @@
-        self.unit.status = ActiveStatus()
+        self._alertmanager_configurer_container_name = (
+            self._alertmanager_configurer_layer_name
+        ) = self._alertmanager_configurer_service_name = self.ALERTMANAGER_CONFIGURER_SERVICE_NAME
+        self._dummy_http_server_container_name = (
+            self._dummy_http_server_layer_name
+        ) = self._dummy_http_server_service_name = self.DUMMY_HTTP_SERVER_SERVICE_NAME
+        self._alertmanager_configurer_container = self.unit.get_container(
+            self._alertmanager_configurer_container_name
+        )
+        self._dummy_http_server_container = self.unit.get_container(
+            self._dummy_http_server_container_name
+        )
+
+        self.service_patch = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="alertmanager-config", port=self.ALERTMANAGER_CONFIGURER_PORT),
+                ServicePort(name="dummy-http-server", port=self.DUMMY_HTTP_SERVER_PORT),
+            ],
+        )
+
+        self.framework.observe(
+            self.on.alertmanager_configurer_pebble_ready,
+            self._on_alertmanager_configurer_pebble_ready,
+        )
+        self.framework.observe(
+            self.on.dummy_http_server_pebble_ready, self._on_dummy_http_server_pebble_ready
+        )
+        self.framework.observe(
+            self.on.alertmanager_relation_joined, self._on_alertmanager_relation_joined
+        )
+        self.framework.observe(self.on.config_changed, self._on_alertmanager_config_changed)
+        self.framework.observe(
+            self.on.alertmanager_config_file_changed, self._on_alertmanager_config_changed
+        )
+        self.framework.observe(
+            self.on.alertmanager_configurer_relation_joined,
+            self._on_alertmanager_configurer_relation_joined,
+        )
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+    def _on_alertmanager_configurer_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Checks whether all conditions to start Alertmanager Configurer are met and, if yes,
+        triggers start of the alertmanager-configurer service.
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
+        Args:
+            event: Juju PebbleReadyEvent event
 
-        Learn more about config at https://juju.is/docs/sdk/config
+        Returns:
+            None
         """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+        watchdog = AlertmanagerConfigDirWatcher(self, self.ALERTMANAGER_CONFIG_DIR)
+        watchdog.start_watchdog()
```

Comment:
Won't this start a bunch of watchdogs? Especially if the `pebble_ready` is deferred a couple of times?

Reply:
True.
Fixed.